### PR TITLE
FIX: return the actual ax.get_window_extent

### LINKED ISSUE #12697: Axes are drawn at wrong positions
### Bug report

**Bug summary**

The window extent of the axes does not coincide with their position in the figure. This issue occurs in Agg and Cairo backends.

**Code for reproduction**

The following code produces two axes. A green rectangle is drawn at the bounding box of the axes. A red rectangle is drawn in the space between the axes.

```
import matplotlib.pyplot as plt

fig, axs = plt.subplots(2, figsize=(4,2), dpi=100)
for ax in axs.flatten():
    ax.set(xticklabels = [], yticklabels=[])
    ax.tick_params(labelbottom=False, bottom=False)

fig.canvas.draw()

bb1 = axs[0].get_window_extent()
bb2 = axs[1].get_window_extent()

rect = plt.Rectangle((bb2.x0, bb2.y1), width=20, height=bb1.y0 - bb2.y1,
                     linewidth=72./fig.dpi, edgecolor="red", facecolor="none", 
                     transform=None)
rect2 = plt.Rectangle((bb2.x0, bb2.y0), width=bb2.width, height=bb2.height,
                     linewidth=72./fig.dpi, edgecolor="limegreen", facecolor="none", 
                     transform=None)
fig.add_artist(rect2)
fig.add_artist(rect)

fig.savefig("borders.png")
plt.show()
```

**Actual outcome**

Code run with 100 dpi
borders

Code run with 200 dpi
borders

The green rectangle does not overlay with the axes. The red rectangle does not sit tight on both axes.

**Expected outcome**

The green rectangle should appear exactly on top of the axes spines. The red rectangle should sit tight on both axes.

**Matplotlib version**

  * Operating system: Win8
  * Matplotlib version: master
  * Matplotlib backend: Qt5Agg, TkCairo
  * Python version: 3.6.6

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

Closes #12697 

Previously, `ax.get_window_extent()` tried to take into account ticks around the axes.  It did this incorrectly because it did not take into account dpi of the canvas, and it just put a pad around the whole axes, without regard for where the ticks actually were.  

This PR changes  `ax.get_window_extent()` so that it does *not* have a pad around the outside, and hence just returns a bbox that is fit to the size of the axes.  

It moves the tick padding  logic into `spine.get_window_extent()` where it can check tick direction, and pad the spine bbox appropriately.  

It also fixes `ax.axison` being ignored by `ax.get_tightbbox`; if the x/yaxis isn't visible, then it should not be part of the tightbbox.  

 

```python
import matplotlib.pyplot as plt

fig, axs = plt.subplots(1, figsize=(4,2))
#for ax in axs.flatten():
    # ax.set(xticklabels = [], yticklabels=[])
    # ax.tick_params(labelbottom=False, bottom=False)

fig.canvas.draw()

bb2 = axs.get_window_extent()

rect2 = plt.Rectangle((bb2.x0, bb2.y0), width=bb2.width, height=bb2.height,
                     linewidth=72./fig.dpi, edgecolor="limegreen", facecolor="none",
                     transform=None)
fig.add_artist(rect2)
fig.savefig('TestExtent.png', dpi=200)
plt.show()

```

### Before:

Extents are slightly larger because they are supposed to account for the ticks.  But the tick size was not properly scaled into pixels

testextentold

### Now:

testextent

## More illustrative test

```python
import matplotlib.pyplot as plt

fig, axs = plt.subplots(2, 2, figsize=(6,6))
fig.subplots_adjust(bottom=0.15, left=0.15, hspace=0.4, wspace=0.4)
axs = axs.flat
axs[1].set(xticklabels = [], yticklabels=[])
axs[1].tick_params(labelbottom=False, bottom=False)

ax = axs[2]
ax.spines['left'].set_position('center')
ax.spines['right'].set_visible(False)
ax.spines['bottom'].set_position(('axes', -0.1))
ax.spines['top'].set_visible(False)
ax.xaxis.set_ticks_position('bottom')
ax.yaxis.set_ticks_position('left')

ax = axs[3]
ax.spines['left'].set_position('center')
ax.spines['right'].set_visible(False)
ax.spines['bottom'].set_position(('axes', -0.1))
ax.spines['top'].set_visible(False)
ax.xaxis.set_ticks_position('bottom')
ax.yaxis.set_ticks_position('left')
axs[3].set(xticklabels = [], yticklabels=[])
axs[3].tick_params(labelbottom=False, bottom=False)

def color_boxes(fig, axs):
    fig.canvas.draw()
    renderer = fig.canvas.get_renderer()
    for axx in [axs.xaxis, axs.yaxis]:
        bb = axx.get_tightbbox(renderer)
        if bb:
            axisr = plt.Rectangle((bb.x0, bb.y0), width=bb.width, height=bb.height,
                             linewidth=0.7, edgecolor='y', facecolor="none",
                             transform=None, zorder=3)
            fig.add_artist(axisr)
    for nn, a in enumerate(['bottom', 'top', 'left', 'right']):
        bb = axs.spines[a].get_window_extent(renderer)
        if bb:
            spiner = plt.Rectangle((bb.x0, bb.y0), width=bb.width, height=bb.height,
                             linewidth=0.7, edgecolor=f'C{nn}', facecolor="none",
                             transform=None, zorder=3)
            fig.add_artist(spiner)

    bb2 = axs.get_window_extent()
    rect2 = plt.Rectangle((bb2.x0, bb2.y0), width=bb2.width, height=bb2.height,
                         linewidth=1.5, edgecolor="magenta", facecolor="none",
                         transform=None, zorder=2)
    fig.add_artist(rect2)

    bb2 = axs.get_tightbbox(renderer)
    rect2 = plt.Rectangle((bb2.x0, bb2.y0), width=bb2.width, height=bb2.height,
                         linewidth=3, edgecolor="r", facecolor="none",
                         transform=None, zorder=1)
    fig.add_artist(rect2)


for ax in axs:
    color_boxes(fig, ax)

```

### Key:

 - thick red: `ax.get_tightbbox()`
 - thick magenta: `ax.get_window_extent()`
 - thin yellow: `ax.yaxis.get_tightbbox()`, `ax.xaxis.get_tightbbox()`, 
 - thin colors: `ax.spines.get_window_extent()`.

### Before

- note how the tightbbox does not cover the ticks (the original bug report)

testextentold

### After:

- Note that spine bboxes now include the ticks, whereas before it was just the spines.  This is a change.

testextentold

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by anntzer** [comment]
The x-ticks are still not in?

**Comment by jklymak** [comment]
oops, sorry, old PNG...

**Comment by jklymak** [comment]
BTW, I'm open to the ticks *not* being included in `ax.get_window_extent` and moving them to be part of the `ax.get_tightbbox`.  Its a bit arbitrary where to decide what is part of the intrinsic axes versus the whole axes including all decorators. I have a funny feeling that for #12690 we are going to want to know where the spines are w/o the tick padding.

**Comment by jklymak** [comment]
Only 4 tests changed because of this!  No wonder no one knew...

**Comment by pharshalp** [comment]
just my two cents ... it would be better not to include ticks in `ax.get_window_extent` since the result would depend on whether ticks are drawn pointing inward or outward. We could also add an optional argument (something like `ax.get_window_extent(include_ticks=True)` to be more flexible about what is considered to be a part of the axes.

**Comment by jklymak** [comment]
Or we could move ticks to the axis (vs axes) window extent.  Or have a spine extent.

**Comment by ImportanceOfBeingErnest** [comment]
I think the underlying question is: If you specify 

```
plt.figure(figsize=(5,5), dpi=100)
fig.subplot_adjust(left=0.1)
```

do you then expect the tick ends to be 50 pixels away from the left figure edge or the axes spine?  
My vote here would clearly go to the axes spine.

However, for the tight_layout calculations, one would probably not expect to have overlapping ticks. So here the tick ends should be relevant. 

In total this might make things a bit more complicated?

<hr>

> Only 4 tests changed because of this!

Isn't that quite worrying? I would expect *all*  image tests to fail. If they don't, it would mean that the tests do not test the actual figures being produced. So is this backend dependent? Or is there an rc setting that steers this?

**Comment by ImportanceOfBeingErnest** [comment]
Yes, I confused myself. The axes are at the correct positions and it's really `get_window_extent` that is wrong. Also I fell again for #9913, which makes it look more wrong on screen than in the saved image.

File: lib/matplotlib/axes/_base.py [hunk 1]
```diff
@@ -4322,13 +4321,14 @@ def get_tightbbox(self, renderer, call_axes_locator=True,
         else:
             self.apply_aspect()
 
-        bb_xaxis = self.xaxis.get_tightbbox(renderer)
-        if bb_xaxis:
-            bb.append(bb_xaxis)
+        if self.axison:
```
Comment (jklymak) [inline lib/matplotlib/axes/_base.py]: This change (self.axison), caused a few tests top change.  So this is probably a breaking change.  OTOH, I think its correct - it doesn't make sense to include the axis in the tight layout if its been turned off.

**Comment by tacaswell** [comment]
:+1: in principle.  Agree with the behavior changes.

File: doc/api/next_api_changes/2018-11-04-JMK.rst [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+that was slightly larger than the axes, presumably to take into account
+the ticks that may be on a spine.  However, it was not scaling the tick sizes
+according to the dpi of the canvas, and it did not check if the ticks were
+visible, or on the spine.
+
+Now  `.matplotlib.axes.Axes.get_window_extent` just returns the axes extent
+with no padding for ticks.
+
+`.spines.get_window_extent` now takes into account ticks that are on the
+spine.
```
Comment (anntzer) [inline doc/api/next_api_changes/2018-11-04-JMK.rst]: Is a note needed for get_tightbbox changes?  (Haven't really followed whether it changed)

**Comment by anntzer** [comment]
It's not totally clear to me why the tick is part of the spine, but the label is not?
For me the grouping is {spine, {tick, label}} (for example the ticker machinery handles ticks and labels), whereas the grouping proposed here is more {{spine, tick}, label}.

**Comment by jklymak** [comment]
> It's not totally clear to me why the tick is part of the spine, but the label is not?
> For me the grouping is {spine, {tick, label}} (for example the ticker machinery handles ticks and labels), whereas the grouping proposed here is more {{spine, tick}, label}.

Well, I'm not proposing to change the existing organization, which I won't pretend to understand.  I'm just making `ax.spines.get_window_extent()` and hence `ax.get_window_extent()` fully enclose stuff we may put on the page, and not over-enclose blank areas.  How that all gets organized inside `ax.get_window_extent()` is a different battle.

**Comment by anntzer** [comment]
Right now you changed spine.get_extent to include the ticks; I'm proposing to change axis.get_extent to do that instead.  (Following a very cursory read...)

**Comment by jklymak** [comment]
Can ticks be detached from the spine?  I thonk placing the labels depends on knowing where the spine extent, including ticks, so I can at least see a case for separating these two things.

File: doc/api/next_api_changes/2018-11-04-JMK.rst (hunk 2 above)
Comment (jklymak) [reply doc/api/next_api_changes/2018-11-04-JMK.rst]: Agreed, it changes if the ticks are outwards and there are no outwards tick labels.

**Comment by anntzer** [comment]
(I need more time to think about the logical grouping of ticks and labels question before approving this, but if anyone else wants to have a go at the review, go for it.)

**Comment by NelleV** [comment]
@jklymak Why are you creating a new test file? (lib/matplotlib/tests/test_get_extents.py)
That doesn't follow Python conventions, and while I realize that the axes module is absolutely humongous (and should probably eventually be splitted up in smaller chunks), I don't think it is a good idea to not follow conventions. It means that it is likely that contributors are not going to realize the tests are in this file, and we may end up with duplicate tests.

**Comment by jklymak** [comment]
> (I need more time to think about the logical grouping of ticks and labels question before approving this, but if anyone else wants to have a go at the review, go for it.)

Just to be clear - currently this is the organization that already exists, this PR just does it correctly.  If we want a different PR to change the organization, we can certainly do that, but I have no idea of the knock-on effects.  

> Why are you creating a new test file? (lib/matplotlib/tests/test_get_extents.py)

Happy to change this.  OTOH note that this tests methods from `spines`, `axis`, and `_axes`, and I didn't think it was particularly helpful to split them all up...

**Comment by jklymak** [comment]
Marking as release critical because this has sat around for a long time.  The only outstanding issue is where the tests should go, but I don't actually see the problem with adding a new file.

**Comment by jklymak** [comment]
ping @ImportanceOfBeingErnest for a review - this closes a bug you reported!

**Comment by ImportanceOfBeingErnest** [comment]
So this fixes the original bug. :+1: 

<hr>


As far as I can see there is no drawback of organizing things like they are here, i.e. have `ax.get_window_extent()` return the axes position, and `ax.get_tightbox()` return the union of all the stuff that's around the axes. Except that maybe the names of the functions would suggest it to be just the other way around? Anyways, I'm absolutely not proposing to change that. But explicitely mentionning in the docstring of `Axes.get_window_extent` something like 

> "This explicitely does not include ticks, labels and spines. Those would be part of the bbox returned by `ax.get_tightbbox` instead. 

might help people using or working on this.

<hr>

Would 

    ax.get_position().transformed(fig.transFigure).bounds == ax.get_window_extent().bounds

be always true? I think so. Could be an additional test?

@NelleV Do you have an alternative suggestion on how to organize the tests? Would you add them all to `test_axes.py` even though they test spines and ax**i**s as well. Should they be splitted up? Would it be enough to add a comment in the respective files like

> `# Tests for the window_extent and tightbbox of axes, axis and spines are`   
> `# found in test_get_extents.py`

<hr>

Finally, I wanted to mention that this PR leads to the fact that ticks or labels that are not shown still have a window_extent:

image

To be clear, this doesn't hurt any of the layout managers, so I'm personally fine with that. Just looks strange and might be confusing to someone in the future.

**Comment by jklymak** [comment]
Thanks.  The last point is interesting. I’m fine with pie behaving that way - it’s the same as for any axes that sets its aspect ratio.  But axis off looks to just make the elements invisible but still taking space whereas I’d have thought it would get rid of them.

**Comment by jklymak** [comment]
documentation change handled.

**Comment by jklymak** [comment]
`self.bbox` is given by 
```
self.bbox = mtransforms.TransformedBbox(self._position,
                                                fig.transFigure)
```
so for sure `ax.get_position().transformed(fig.transFigure).bounds == ax.get_window_extent().bounds` is true...

**Comment by jklymak** [comment]
Extra test added to make sure the roundtrip works.

**Comment by NelleV** [comment]
Because the standard in Python is to have a test file per module, the problem with the organization of oure tests is actually due to the size of the axes module: it's way too big. We already splitted in two in 2013, and we talked about continuing to refactor it to have smaller size objects and modules but never went forward with it.

**Comment by tacaswell** [comment]
@jklymak needs another re-base.  I think those tests no longer exist?

**Comment by jklymak** [comment]
Fixed - not the outward ticks test changed a bit, and I restructured it so that I could get the right numbers when the comparison failed.  I could put it back, but I actually think this is more readable...

**Comment by dstansby** [comment]
Woops, just accidentally pressed some reveiw request buttons, feel free to ignore...

**Comment by tacaswell** [comment]
The appevyor failure was upload related.

File: lib/matplotlib/axes/_base.py [hunk 3]
```diff
@@ -4286,6 +4289,13 @@ def get_tightbbox(self, renderer, call_axes_locator=True,
         -------
         bbox : `.BboxBase`
             bounding box in figure pixel coordinates.
+
+        See Also
+        --------
+        matplotlib.axis.Axes.get_window_extent
```
Comment (bsipocz) [inline lib/matplotlib/axes/_base.py]: These  extra 'See also' sections with non functional API links are breaking more strict docs builds downstream.
https://matplotlib.org/3.1.0/api/_as_gen/matplotlib.axes.Axes.get_tightbbox.html#matplotlib.axes.Axes.get_tightbbox

(We run the docs build in our CI in astropy with the `-w` option, so any sphinx warning is causing a failure (as our WCSAxes class is inherited from `Axes` and ultimately from this docstring).

If it's helpful I can open a separate issue for this.

Comment (QuLogic) [reply lib/matplotlib/axes/_base.py]: A new issue would get better traction.

Comment (tacaswell) [reply lib/matplotlib/axes/_base.py]: We also run sphinx with `-W` in our options https://github.com/matplotlib/matplotlib/blob/512aecb3ba330a46c09f42683ae1e90d8302dcff/doc/Makefile#L5
so it is odd that astropy is failing but we are not...

I suspect difference is sphinx versions (we are currently pinning to sphinx < 2 due to weird spacing at TOC and some breakage (which may be this)).

It is also odd that this is not linking an the 'previous' link in the top right is the thing we want this to have linked to.


Agree with @QuLogic , this should get it's own issue.

**Pull request merged**